ygot/pathstrings.go

```diff
@@ -0,0 +1,268 @@
+// "/one/two/three" and returned. Both the pre-0.4.0 "element"-based paths, and the
+//  >0.4.0 paths based on "elem" are supported. In the case that post-0.4.0 paths are
+// specified, keys that are specified in the path are concatenated onto the name of
+// the path element using the format [name=value]. If the path specifies both pre-
+// and post-0.4.0 paths, the pre-0.4.0 version is returned.
+func PathToString(path *gnmipb.Path) (string, error) {
+	p := []string{"/"}
+	if path.Element != nil {
+
+		for i, e := range path.Element {
+			if e == "" {
+				return "", fmt.Errorf("nil element at index %d in %v", i, path.Element)
+			}
+			p = append(p, e)
+		}
+		return filepath.Join(p...), nil
+	}
+
+	for i, e := range path.Elem {
+		if e.Name == "" {
+			return "", fmt.Errorf("nil name for PathElem at index %d", i)
+		}
+
+		elem := e.Name
+		if len(e.Key) != 0 {
+			var keys []string
+			for k, v := range e.Key {
+				if k == "" {
+					return "", fmt.Errorf("nil key name (value: %s) in element %s", v, e.Name)
+				}
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+
+			for _, k := range keys {
+				elem = fmt.Sprintf("%s[%s=%s]", elem, k, e.Key[k])
+			}
+		}
+		p = append(p, elem)
+	}
+	return filepath.Join(p...), nil
+}
+
+// PathType is used to indicate a gNMI path type.
+type PathType int64
+
+const (
+	// StructuredPath represents a Path using the structured 'PathElem' message in
+	// the 'elem' field of the gNMI Path message.
+	StructuredPath PathType = iota
+	// StringSlicePath represents a Path using the 'Element' repeated string field
+	// of the gNMI path message.
+	StringSlicePath
+)
+
+// StringToPath takes an input string representing a path in gNMI, and converts
+// it to a gNMI Path message, populated with the specified path encodings.
+func StringToPath(path string, pathTypes ...PathType) (*gnmipb.Path, error) {
+	var errs util.Errors
+	if len(pathTypes) == 0 {
+		return nil, util.AppendErr(errs, errors.New("no path types specified"))
+	}
+
+	pmsg := &gnmipb.Path{}
+	for _, p := range pathTypes {
+		switch p {
+		case StructuredPath:
+			gp, err := StringToStructuredPath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building structured path: %v", err))
+				continue
+			}
+			pmsg.Elem = gp.Elem
+		case StringSlicePath:
+			gp, err := StringToStringSlicePath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building string slice path: %v", err))
+				continue
+			}
+			pmsg.Element = gp.Element
+		}
+	}
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	return pmsg, nil
+}
+
+// StringToStringSlicePath takes a string representing a path, and converts it into a
+// gnmi.Path. For example, if the Path "/a/b[c=d]/e" is input, it is converted
+// to a gnmi.Path{Element: []string{"a", "b[c=d]", "e"}} which is returned.
+func StringToStringSlicePath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+	for _, p := range parts {
```

Comment:
Is it possible to put this check inside pathStringToElements and return error there?

Reply:
It's possible to, yes, but it means that if `pathStringToElements` is run then `extractKV` (as is done below) then you range the string three times rather than two that we have currently.